substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+use frame_election_provider_support::SortedListProvider;
+use frame_support::{
+	defensive,
+	traits::{fungible::Inspect as FnInspect, Defensive, DefensiveSaturating},
+};
+use sp_runtime::traits::Zero;
+use sp_staking::{
+	currency_to_vote::CurrencyToVote, OnStakingUpdate, Stake, StakerStatus, StakingInterface,
+};
+use sp_std::{collections::btree_map::BTreeMap, vec, vec::Vec};
+
+#[cfg(test)]
+pub(crate) mod mock;
+#[cfg(test)]
+mod tests;
+
+pub(crate) const LOG_TARGET: &str = "runtime::stake-tracker";
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: $crate::LOG_TARGET,
+			concat!("[{:?}] 📚 ", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> = <<T as Config>::Staking as StakingInterface>::Balance;
+/// The account ID of this pallet.
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+/// Represents a stake imbalance to be applied to a staker's score.
+#[derive(Copy, Clone, Debug)]
+pub enum StakeImbalance<Balance> {
+	/// Represents the reduction of stake by `Balance`.
+	Negative(Balance),
+	/// Represents the increase of stake by `Balance`.
+	Positive(Balance),
+}
+
+impl<Balance: PartialOrd + DefensiveSaturating> StakeImbalance<Balance> {
+	fn from(prev_balance: Balance, new_balance: Balance) -> Self {
+		if prev_balance > new_balance {
+			StakeImbalance::Negative(prev_balance.defensive_saturating_sub(new_balance))
+		} else {
+			StakeImbalance::Positive(new_balance.defensive_saturating_sub(prev_balance))
+		}
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::{ExtendedBalance, VoteWeight};
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::BlockNumberFor;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The stake balance.
+		type Currency: FnInspect<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The staking interface.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// Something that provides an *always* sorted list of voters.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+
+		/// Something that provides an *always* sorted list of targets by their approval stake.
+		type TargetList: SortedListProvider<
+			Self::AccountId,
+			Score = <Self::Staking as StakingInterface>::Balance,
+		>;
+	}
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
```

Comment:
```suggestion
```

Reply:
Addressed it here https://github.com/paritytech/polkadot-sdk/pull/4411